pkg/storage/pebbleiter/crdb_test_off.go

```diff
@@ -0,0 +1,26 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+//go:build !crdb_test || crdb_test_off
```

Comment:
Re: the last bit -- if we're going to keep track of whether each individual iterator is closed to warn about double-closes, I also wonder whether that logic makes more sense as a part of the `pebble.Iterator` itself. That struct could carry around the `closed bool` and give a sensible error if the method is called more than once.

Reply:
> However I wonder if the extra cost (another little struct and maybe one additional indirection for Clone() and Close()) is not minimal enough that we can just eat it. I don't know enough about the implementation to say so -- maybe these iterators are opened and closed in a really hot loop and we can't afford taking any extra time.

This layer is particularly performance sensitive and any additional allocations would be prohibitive.

> Re: the last bit -- if we're going to keep track of whether each individual iterator is closed to warn about double-closes, I also wonder whether that logic makes more sense as a part of the pebble.Iterator itself. That struct could carry around the closed bool and give a sensible error if the method is called more than once.

Since both `storage.pebbleIterator` and `pebble.Iterator` are pooled to reduce allocations, the iterators structs may already be in use again by the time `Close` is called a second time. We can't distinguish between a `Close` issued by the new user, versus a double-`Close` by the old user.

It does seem like it'd be feasible to maintain monotonically increasing  counts of `opens` and `closes` and assert that `closes` never exceeds `opens`, but the ability to make `crdb_test`-gated assertions of `pkg/storage`'s use of `pebble.Iterator` is generally useful. We could move some of those assertions into `pebble.Iterator`, but they still need to be gated on `crdb_test`.

@rickystewart would it be feasible to add the Pebble `invariants` build tag when the Cockroach `crdb_test` build tag is set? That could allow us to take advantage of the existing Pebble invariant assertions.

Reply:
(As mentioned, don't let this comment block you, this is fine to be checked-in as is.)

> @rickystewart would it be feasible to add the Pebble invariants build tag when the Cockroach crdb_test build tag is set? That could allow us to take advantage of the existing Pebble invariant assertions.

It is possible, but it might result in a lot of stuff being re-compiled for developers when swapping between builds and tests if they happen to be using `pebble`. You can file a follow-up issue to discuss this if you like.